File: reference/fx870_es/Trace.pas
{ Execution trace unit.
  Writes one JSONL record per instruction to trace.jsonl.
  Recording starts when TraceInit is called and stops automatically after
  TRACE_DURATION_MS milliseconds (wall clock) or TRACE_MAX_LINES records,
  whichever comes first.  TraceClose can also be called explicitly.

  Hook in cpu.pas:
    - call TraceInstr    immediately before ExecInstr
    - call TraceAddCycles(cycles) after the speed-scaling line in CpuRun

  Hook in main.pas:
    - call TraceInit  after ResetAll in FormShow
    - call TraceClose before IoClose in FormClose
}

unit Trace;

interface

  procedure TraceInit;
  procedure TraceClose;
  procedure TraceInstr;
  procedure TraceAddCycles (n: integer);

implementation

uses
  SysUtils, Windows, Def;

const
  TRACE_DURATION_MS = 10000;   { wall-clock limit in milliseconds }
  TRACE_MAX_LINES   = 5000000; { hard line-count safety cap }
  TRACE_FILE        = 'trace.jsonl';

var
  TraceFile     : TextFile;
  TraceOpen     : boolean = False;
  TraceStartTick: DWORD;
  TraceCycles   : int64;
  TraceLines    : integer;


{ Serialise mr[0..35] as a compact JSON integer array, e.g. [0,1,255,...] }
function MrJson: string;
var
  i: integer;
  s: string;
begin
  s := '[';
  for i := 0 to 35 do
  begin
    if i > 0 then s := s + ',';
    s := s + IntToStr(mr[i]);
  end {for};
  Result := s + ']';
end {MrJson};


{ Open (or re-open) the trace file and begin recording. }
procedure TraceInit;
begin
  if TraceOpen then CloseFile(TraceFile);
  AssignFile(TraceFile, TRACE_FILE);
  Rewrite(TraceFile);
  TraceOpen     := True;
  TraceStartTick := GetTickCount;
  TraceCycles   := 0;
  TraceLines    := 0;
end {TraceInit};


{ Flush and close the trace file. }
procedure TraceClose;
begin
  if not TraceOpen then Exit;
  CloseFile(TraceFile);
  TraceOpen := False;
end {TraceClose};


{ Called BEFORE ExecInstr in CpuRun.
  At this point pc holds the address of the instruction about to execute.
  TraceCycles is the cumulative cycle count up to (but not including) this
  instruction - it is updated afterwards via TraceAddCycles. }
procedure TraceInstr;
var
  line: string;
begin
  if not TraceOpen then Exit;

  { check wall-clock timeout }
  if (GetTickCount - TraceStartTick) >= TRACE_DURATION_MS then
  begin
    TraceClose;
    Exit;
  end {if};

  { check line-count cap }
  if TraceLines >= TRACE_MAX_LINES then
  begin
    TraceClose;
    Exit;
  end {if};

  line := Format(
    '{"pc":%d,"cy":%d,"mr":%s,' +
    '"sx":%d,"sy":%d,"sz":%d,' +
    '"ix":%d,"iy":%d,"iz":%d,' +
    '"us":%d,"ss":%d,' +
    '"ua":%d,"dua":%d,' +
    '"ia":%d,"ie":%d,"ib":%d,' +
    '"tm":%d,"fl":%d,"iserv":%d,"ky":%d}',
    [pc,   TraceCycles, MrJson,
     sx,   sy,          sz,
     ix,   iy,          iz,
     us,   ss,
     ua,   delayed_ua,
     ia,   ie,           ib,
     tm,   flag,         iserv,  ky]);

  WriteLn(TraceFile, line);
  Inc(TraceLines);
end {TraceInstr};


{ Accumulate the cycle count returned by CpuRun so that TraceCycles stays
  synchronised with the emulated clock. }
procedure TraceAddCycles (n: integer);
begin
  if TraceOpen then Inc(TraceCycles, n);
end {TraceAddCycles};


end.

File: src/emulator/trace.ts
// Execution trace module — mirrors Delphi Trace.pas
// Records one JSONL line per instruction to /trace (→ trace.jsonl on disk).
// Output format matches Trace.pas so the two traces can be diffed directly.
//
// Usage:
//   traceInit()          — start recording (call from emulatorReset)
//   traceClose()         — flush & stop (auto-fires on cycle/line limit)
//   traceInstr(pc)       — call in cpuRun() before execInstr()
//   traceAddCycles(n)    — call in cpuRun() after the speed-scale line
//   isTraceEnabled()     — guard used in cpu.ts hot path
//
// TRACE_MAX_CYCLES is set to 10 s of nominal 921 kHz wall time.
// The server plugin in vite.config.ts appends batches to trace.jsonl.

import {
  mr,
  sx, sy, sz,
  ix, iy, iz,
  us, ss,
  ua, delayed_ua,
  ia, ie, ib,
  tm, flag, iserv, ky,
} from './def.js';

// 10 s at 921 kHz ≈ 9 210 000 cycles
const TRACE_MAX_CYCLES = 921 * 10_000;
const TRACE_MAX_LINES  = 5_000_000;
const BATCH_SIZE       = 500; // lines per POST

let _enabled   = false;
let _cycles    = 0;           // cumulative emulated cycles
let _lines     = 0;
let _batch: string[] = [];

export function traceInit(): void {
  _enabled = true;
  _cycles  = 0;
  _lines   = 0;
  _batch   = [];
}

export function traceClose(): void {
  if (!_enabled) return;
  _flush();
  _enabled = false;
}

export function isTraceEnabled(): boolean {
  return _enabled;
}

// Called from cpuRun() BEFORE execInstr().
// pc is the address of the instruction about to execute.
export function traceInstr(pc: number): void {
  if (!_enabled) return;

  if (_cycles >= TRACE_MAX_CYCLES || _lines >= TRACE_MAX_LINES) {
    traceClose();
    return;
  }

  // Build compact JSON matching Trace.pas field names
  const mrArr = '[' + (mr as Uint8Array).join(',') + ']';
  _batch.push(
    `{"pc":${pc},"cy":${_cycles},"mr":${mrArr},` +
    `"sx":${sx},"sy":${sy},"sz":${sz},` +
    `"ix":${ix},"iy":${iy},"iz":${iz},` +
    `"us":${us},"ss":${ss},` +
    `"ua":${ua},"dua":${delayed_ua},` +
    `"ia":${ia},"ie":${ie},"ib":${ib},` +
    `"tm":${tm},"fl":${flag},"iserv":${iserv},"ky":${ky}}`
  );
  _lines++;

  if (_batch.length >= BATCH_SIZE) _flush();
}

// Called from cpuRun() after the speed-scale line, with the final cycle count.
export function traceAddCycles(n: number): void {
  if (_enabled) _cycles += n;
}

// Serial send queue: ensures batches arrive at the server in order.
let _sendQueue: Promise<void> = Promise.resolve();

function _flush(): void {
  if (_batch.length === 0) return;
  const body = _batch.join('\n') + '\n';
  _batch = [];
  _sendQueue = _sendQueue.then(() =>
    fetch('/trace', {
      method: 'POST',
      headers: { 'Content-Type': 'text/plain' },
      body,
    }).then(() => {}).catch(() => { /* ignore if dev server not running */ })
  );
}
